### PR TITLE
ARROW-728: [C++/Python] Add Table::RemoveColumn method, remove name member, some other code cleaning

### PR DESCRIPTION
**Pull request by wesm**
* Consolidated column.h and table.h
* Consolidated schema.h and type.h
* Removed some `Equals(const std::shared_ptr<T>&)` methods, better to use `const T&` methods

### CODE REVIEW

**Comment by wesm** [comment]
This build fails because the API change impacts Parquet. @xhochy we'll need to merge this in broken state and then update the version in https://github.com/apache/parquet-cpp/pull/280 to get a green build

File: cpp/src/arrow/util/stl.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#define ARROW_UTIL_STL_H
+
+#include <vector>
+
+namespace arrow {
+
+template <typename T>
+void DeleteVectorElement(
+    const std::vector<T>& values, size_t index, std::vector<T>* out) {
+  out->reserve(values.size() - 1);
```
Comment (xhochy) [inline cpp/src/arrow/util/stl.h]: Why not return the new vector here? This function has no status.

**Review by xhochy — APPROVED** [review]
Lgtm. Feel free to merge.

File: cpp/src/arrow/util/stl.h (hunk 1 above)
Comment (wesm) [inline cpp/src/arrow/util/stl.h]: I guess I should be more trusting of RVO. Will do that

**Comment by wesm** [comment]
OK, I think we're just down to Parquet failures now. Merging

**Pull request closed without merging**